src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/NumericIndexWriter.java

```diff
@@ -0,0 +1,179 @@
+                              int bytesPerValue,
+                              long maxSegmentRowId)
+    {
+        checkArgument(maxSegmentRowId >= 0, "[%s] maxRowId must be non-negative value, but got %s", indexContext.getIndexName(), maxSegmentRowId);
+
+        this.indexDescriptor = indexDescriptor;
+        this.indexContext = indexContext;
+        this.bytesPerValue = bytesPerValue;
+        this.writer = new BlockBalancedTreeWriter(maxSegmentRowId + 1, bytesPerValue, maxPointsInLeafNode);
+    }
+
+    @Override
+    public String toString()
+    {
+        return MoreObjects.toStringHelper(this)
+                          .add("bytesPerDim", bytesPerValue)
+                          .add("bufferedPoints", writer.getPointCount())
+                          .toString();
+    }
+
+    public static class LeafCallback implements BlockBalancedTreeWriter.Callback
+    {
+        final List<PackedLongValues> postings = new ArrayList<>();
+
+        public int numLeaves()
+        {
+            return postings.size();
+        }
+
+        @Override
+        public void writeLeafDocs(BlockBalancedTreeWriter.RowIDAndIndex[] sortedByRowID, int offset, int count)
+        {
+            PackedLongValues.Builder builder = PackedLongValues.monotonicBuilder(PackedInts.COMPACT);
+
+            for (int i = offset; i < count; ++i)
+            {
+                builder.add(sortedByRowID[i].rowID);
+            }
+            postings.add(builder.build());
+        }
+    }
+
+    /**
+     * Writes a balanced tree and posting lists from a {@link org.apache.lucene.codecs.MutablePointValues}.
+     *
+     * @param values points to write
+     *
+     * @return metadata describing the location and size of this balanced tree in the overall SSTable balanced tree component file
+     */
+    public SegmentMetadata.ComponentMetadataMap writeAll(IntersectingPointValues values) throws IOException
+    {
+        long balancedTreePosition;
+        SegmentMetadata.ComponentMetadataMap components = new SegmentMetadata.ComponentMetadataMap();
+
+        LeafCallback leafCallback = new LeafCallback();
+
+        try (IndexOutput balancedTreeOutput = indexDescriptor.openPerIndexOutput(IndexComponent.BALANCED_TREE, indexContext, true))
+        {
+            // The SSTable balanced tree component file is opened in append mode, so our offset is the current file pointer.
+            long balancedTreeOffset = balancedTreeOutput.getFilePointer();
+
+            balancedTreePosition = writer.writeField(balancedTreeOutput, values, leafCallback);
+
+            // If the balancedTreePosition is less than 0 then we didn't write any values out
+            // and the index is empty
+            if (balancedTreePosition < 0)
+                return components;
+
+            long balancedTreeLength = balancedTreeOutput.getFilePointer() - balancedTreeOffset;
+
+            Map<String, String> attributes = new LinkedHashMap<>();
```

Comment:
nit: Also applies to the attributes for postings below, but do we actually ever do anything with these attributes? They're cheap, but just wondering if we have even a speculative use for them...

Reply:
I think I answered this earlier - we don't currently do anything with them but I think it's worth leaving them there because we could create offline tooling that made use of them for index analysis.